### PR TITLE
UCP: ucp_tag_send_nbr() definition

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe 
This function provides a convenient and efficient way to implement
a blocking send pattern ex MPI_send()

For medium size messages it completes requests faster than
ucp_tag_send_nb() because:
- it always uses bcopy to send up to the rndv threshold
- its rndv threshold is higher than the one used by ucp_tag_send_nb()
- its request handling is simpler: no callback, and no need to
allocate/free. request can be allocated by caller on the stack.

### CODE REVIEW

**Comment by yosefe** [comment]
@shamisp @bbenton pls take a look

**Comment by shamisp** [comment]
I think the justification for the API has to part of the API. Essentially compare the call to ucp_tag_send_nb

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5788/ for details (Mellanox internal link).

**Comment by alex-mikheev** [comment]
@shamisp it is a part of the function description in the doxygen.

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5794/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5799/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -2049,6 +2049,76 @@ ucs_status_ptr_t ucp_tag_send_nb(ucp_ep_h ep, const void *buffer, size_t count,
...
+ * by the UCX_SEND_NBR_RNDV_THRESH environment variable.
+ *
+ * @note This routine provides a convinient and efficient way to implement
+ *       a blocking send pattern. For example MPI_send() pseudo code may
+ *       look like this:
+ *
+ * @code
+ * MPI_send(...)
+ * {
+ *     char request[ucp_request_size];
```
Comment (evgeny-leksikov) [inline src/ucp/api/ucp.h]: Should it be `sizeof(ucp_request_t)`? if so, AFAIR `ucp_request_t` is not an API type.

File: src/ucp/api/ucp.h [hunk 2]
```diff
@@ -2049,6 +2049,76 @@ ucs_status_ptr_t ucp_tag_send_nb(ucp_ep_h ep, const void *buffer, size_t count,
...
+ *     ucs_status_t status;
+ *
+ *     status = ucp_tag_send_nbr(ep, ..., request + ucp_request_size);
+ *     if (status != UCS_INPROGRESS) {
+ *         return status;
+ *     }
+ *
+ *     do {
+ *         ucp_worker_progress(worker);
+ *         status = ucp_request_check_status(request + ucp_request_size);
```
Comment (evgeny-leksikov) [inline src/ucp/api/ucp.h]: This trick is based on internal implementation, looks dangerous.

File: src/ucp/api/ucp.h (hunk 1 above)
Comment (alex-mikheev) [reply src/ucp/api/ucp.h]: no, it is a request size as reported by ucp_context_query()

File: src/ucp/api/ucp.h (hunk 2 above)
Comment (alex-mikheev) [reply src/ucp/api/ucp.h]: We already do things this way in ucp_tag_recv_nbr()

File: src/ucp/api/ucp.h (hunk 1 above)
Comment (evgeny-leksikov) [reply src/ucp/api/ucp.h]: Ok, sorry missed that.

File: src/ucp/api/ucp.h (hunk 2 above)
Comment (evgeny-leksikov) [reply src/ucp/api/ucp.h]: Hmm.. I don't see this hack in ucp_tag_recv_nbr() docs.

Comment (evgeny-leksikov) [reply src/ucp/api/ucp.h]: Maybe better to add smth like this to API?
```
ucs_status ucp_nbr_check_status(void *user_request)
{
    return ucp_request_check_status(user_request + ucp_request_size);
}
```

Comment (alex-mikheev) [reply src/ucp/api/ucp.h]: from ucp_tag_recv_nbr() doc:

>Request handle allocated by the user. There should
> be at least UCP request size bytes of available
 >space before the @a req. The size of UCP request
 > can be obtained by @ref ucp_context_query function.

and
> In order to monitor completion of the operation
 > @ref ucp_request_check_status or @ref ucp_tag_recv_request_test should be
 > used.

Comment (yosefe) [reply src/ucp/api/ucp.h]: @evgeny-leksikov user still has to know UCP request size in order to allocate enough room for it

Comment (evgeny-leksikov) [reply src/ucp/api/ucp.h]: but there is nothing about offset

Comment (yosefe) [reply src/ucp/api/ucp.h]: what do you mean? no documentation?

Comment (evgeny-leksikov) [reply src/ucp/api/ucp.h]: @yosefe yes, but it's very strange to pass pointer to end of allocated memory...

Comment (evgeny-leksikov) [reply src/ucp/api/ucp.h]: yes, following to current docs i would expect to pass pointer to allocated request to check_status_func but not to its end

Comment (alex-mikheev) [reply src/ucp/api/ucp.h]: @evgeny-leksikov ucp_tag_recv_nbr() clearly states that you need to pass a pointers to an end. Requests allocated internally by ucp already return such a pointer to a caller. ex. ucp_tag_send_nb()

Comment (evgeny-leksikov) [reply src/ucp/api/ucp.h]: i see.. thanks for clarifications! but i still dislike this magic :(

**Comment by shamisp** [comment]
@alex-mikheev the difference between the two is not well described in the doxygen. This one is better "its request handling is simpler: no callback, and no need to allocate/free. request can be allocated by caller on the stack." Clarify the differences the best you can. As for now it is not very obvious why we have this API.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5829/ for details (Mellanox internal link).

File: src/ucp/api/ucp.h [hunk 3]
```diff
@@ -2053,6 +2053,17 @@ ucs_status_ptr_t ucp_tag_send_nb(ucp_ep_h ep, const void *buffer, size_t count,
  * @ingroup UCP_COMM
  * @brief Non-blocking tagged-send operations with user provided request
  *
+ * This routine provides a convenient and efficient way to implement a
+ * blocking send pattern. It also completes requests faster than
+ * @ref ucp_tag_send_nbr() because:
+ * @li it always uses bcopy to send data up to the rendezvous threshold.
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: Actually you can link it to uct_bcopy

**Comment by yosefe** [comment]
@shamisp can you pls take a look

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/5836/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
👍

**Comment by yosefe** [comment]
@bbenton  can you pls review as well

**Comment by bbenton** [comment]
:+1: for Doc/API

**Pull request merged**